MediaBrowser.Controller/Library/TaskMethods.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using MediaBrowser.Controller.Configuration;
+
+namespace MediaBrowser.Controller.Library
+{
+    /// <summary>
+    /// Helper methods for running tasks concurrently.
+    /// </summary>
+    public static class TaskMethods
+    {
+        private static readonly int _processorCount = Environment.ProcessorCount;
+
+        private static readonly ConcurrentDictionary<SharedThrottleId, SemaphoreSlim> _sharedThrottlers = new ConcurrentDictionary<SharedThrottleId, SemaphoreSlim>();
+
+        /// <summary>
+        /// Throttle id for sharing a concurrency limit.
+        /// </summary>
+        public enum SharedThrottleId
+        {
+            /// <summary>
+            /// Library scan fan out
+            /// </summary>
+            ScanFanout,
+
+            /// <summary>
+            /// Refresh metadata
+            /// </summary>
+            RefreshMetadata,
+        }
+
+        /// <summary>
+        /// Gets or sets the configuration manager.
+        /// </summary>
+        public static IServerConfigurationManager ConfigurationManager { get; set; }
+
+        /// <summary>
+        /// Similiar to Task.WhenAll but only allows running a certain amount of tasks at the same time.
+        /// </summary>
+        /// <param name="throttleId">The throttle id. Multiple calls to this method with the same throttle id will share a concurrency limit.</param>
+        /// <param name="actions">List of actions to run.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
+        public static async Task WhenAllThrottled(SharedThrottleId throttleId, IEnumerable<Func<Task>> actions, CancellationToken cancellationToken)
+        {
+            var taskThrottler = throttleId == SharedThrottleId.ScanFanout ?
+                new SemaphoreSlim(GetConcurrencyLimit(throttleId)) :
+                _sharedThrottlers.GetOrAdd(throttleId, id => new SemaphoreSlim(GetConcurrencyLimit(id)));
+
+            try
+            {
+                var tasks = new List<Task>();
+
+                foreach (var action in actions)
+                {
+                    await taskThrottler.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                    tasks.Add(Task.Run(async () =>
```

Comment:
While `Task.Run` spawns a new thread, it's not really meant for this kind of use. Spawning a new thread is not all that cheap and the recommendation is to save it for CPU intensive tasks. Metadata refreshes are very network I/O heavy, so most of the time is actually spent waiting.

If you want to achieve parallelism, it's better to use TPL.

Reply:
I switched the subfolder scan to use TPL dataflow - `ActionBlock`s with max `MaxDegreeOfParallelism` set.

However I'm unsure how to limit the amount of concurrent metadata refreshes with TPL. It seems tricky to limit the parallelism since there are multiple independent callers. Any suggestions?